File: PlayPcmWin/PlayPcmWin/RemoteCommand.cs
using System.Collections.Generic;
using System;

namespace PlayPcmWin {
    class RemoteCommandPlayListItem {
        public int durationMillsec;
        public int sampleRate;
        public int bitDepth;
        public string albumName;
        public string artistName;
        public string titleName;
        public byte[] albumCoverArt;
        public RemoteCommandPlayListItem(int aDurationMilliSec,
            int aSampleRate, int aBitDepth, string aAlbumName,
            string aArtistName, string aTitleName, byte[] aAlbumCoverArt) {
            durationMillsec = aDurationMilliSec;
            sampleRate = aSampleRate;
            bitDepth = aBitDepth;
            albumName = aAlbumName;
            artistName = aArtistName;
            titleName = aTitleName;
            albumCoverArt = aAlbumCoverArt;
        }
    };

    enum RemoteCommandType {
        PlaylistData,
        Exit,
        Pause,
        Stop,
        PlayPositionUpdate,

        PPWR,
        PlaylistWant,
        Play,
        Seek,
        SelectTrack,
    };

    class RemoteCommand {
        public const int FOURCC_PLAYLIST_DATA        = 0x534c4c50; // "PLLS" Server -> Client
        public const int FOURCC_EXIT                 = 0x54495845; // "EXIT" both direction
        public const int FOURCC_PAUSE                = 0x53554150; // "PAUS" both direction
        public const int FOURCC_STOP                 = 0x504f5453; // "STOP" Server -> Client
        public const int FOURCC_PLAY_POSITION_UPDATE = 0x55504c50; // "PLPU" Server -> Client

        public const int FOURCC_PPWR                 = 0x52575050; // "PPWR" Client -> Server
        public const int FOURCC_PLAYLIST_WANT        = 0x574c4c50; // "PLLW" Client -> Server
        public const int FOURCC_PLAY                 = 0x59414c50; // "PLAY" Client -> Server
        public const int FOURCC_SEEK                 = 0x4b454553; // "SEEK" Client -> Server
        public const int FOURCC_SELECT_TRACK         = 0x544c4553; // "SELT" Client -> Server

        private static int RemoteCommandTypeToFourCC(RemoteCommandType t) {
            switch (t) {
            case RemoteCommandType.PlaylistData: return FOURCC_PLAYLIST_DATA;
            case RemoteCommandType.Exit: return FOURCC_EXIT;
            case RemoteCommandType.Pause: return FOURCC_PAUSE;
            case RemoteCommandType.Stop: return FOURCC_STOP;
            case RemoteCommandType.PlayPositionUpdate: return FOURCC_PLAY_POSITION_UPDATE;

            default:
                System.Diagnostics.Debug.Assert(false);
                return FOURCC_EXIT;
            }
        }

        // 受信データ用
        private static RemoteCommandType FourCCToRemoteCommandType(int fourcc) {
            switch (fourcc) {
            case FOURCC_EXIT: return RemoteCommandType.Exit;
            case FOURCC_PAUSE: return RemoteCommandType.Pause;
            case FOURCC_PLAYLIST_WANT: return RemoteCommandType.PlaylistWant;
            case FOURCC_PLAY: return RemoteCommandType.Play;
            case FOURCC_SEEK: return RemoteCommandType.Seek;
            case FOURCC_SELECT_TRACK: return RemoteCommandType.SelectTrack;
            default:
                Console.WriteLine("D: Unknown fourcc {0:X8}. defaults to EXIT", fourcc);
                return RemoteCommandType.Exit;
            }
        }

        // 送信用
        public enum PlaybackState {
            Stopped = 0,
            Playing = 1,
            Paused  = 2,
        }

        public RemoteCommandType cmd;
        public int trackIdx;
        public int positionMillisec;
        public PlaybackState state;
        public List<RemoteCommandPlayListItem> playlist = new List<RemoteCommandPlayListItem>();

        public RemoteCommand(RemoteCommandType t, int aTrackIdx=0, int aPositionMs=0) {
            cmd = t;
            trackIdx = aTrackIdx;
            positionMillisec = aPositionMs;
        }

        // 受信したとき用ctor
        public RemoteCommand(int header, int bytes, byte[] payload) {
            cmd = FourCCToRemoteCommandType(header);

            switch (header) {
            case FOURCC_PLAYLIST_WANT:
            case FOURCC_EXIT:
            case FOURCC_PAUSE:
                // no payload
                if (payload.Length != 0) {
                    Console.WriteLine("D: RemoteCommand {0} and payload length is not 0! {1}.", cmd, payload.Length);
                    break;
                }
                break;
            case FOURCC_PLAY:
            case FOURCC_SELECT_TRACK:
                /* trackIdx (4 bytes)
                 */
                if (payload.Length != 4) {
                    Console.WriteLine("D: RemoteCommand {0} and payload length is not 4! {1}.", cmd, payload.Length);
                    break;
                }
                trackIdx = BitConverter.ToInt32(payload, 0);
                if (trackIdx < 0) {
                    Console.WriteLine("D: RemoteCommand {0} and trackIdx is negative value {1}.", cmd, payload.Length);
                    trackIdx = 0;
                    break;
                }
                break;
            case FOURCC_SEEK:
                /* trackIdx (4bytes)
                 * positionMillisec (4bytes)
                 */
                if (payload.Length != 8) {
                    Console.WriteLine("D: RemoteCommand Seek and payload length is not 8 {0}.", payload.Length);
                    break;
                }
                trackIdx = BitConverter.ToInt32(payload, 0);
                if (trackIdx < 0) {
                    Console.WriteLine("D: RemoteCommand Play and trackIdx is negative value {0}.", trackIdx);
                    trackIdx = 0;
                    break;
                }
                positionMillisec = BitConverter.ToInt32(payload, 4);
                if (positionMillisec < 0) {
                    Console.WriteLine("D: RemoteCommand Play and positionMillisec is negative value {0}.", positionMillisec);
                    positionMillisec = 0;
                    break;
                }
                break;
            default:
                System.Diagnostics.Debug.Assert(false);
                break;
            }
        }

        // 送出用バイナリデータを生成する。
        public byte[] GenerateMessage() {
            byte[] r = new byte[0];

            Console.WriteLine("GenerateMessage({0})\n", cmd);

            switch (cmd) {
            case RemoteCommandType.Exit:
            case RemoteCommandType.Stop: {
                // fourcc             (4bytes)
                // payload bytes == 0 (8bytes)
                    r = new byte[12];
                    int fourcc = RemoteCommandTypeToFourCC(cmd);
                    long payloadBytes = 0;
                    Array.Copy(BitConverter.GetBytes(fourcc), 0, r, 0, 4);
                    Array.Copy(BitConverter.GetBytes(payloadBytes), 0, r, 4, 8);
                } break;

            case RemoteCommandType.Pause: {
                    /* fourcc             (4bytes)
                     * payload bytes == 8 (8bytes)
                     * trackIdx           (4bytes)
                     */
                    r = new byte[16];
                    int fourcc = RemoteCommandTypeToFourCC(cmd);
                    long payloadBytes = 4;
                    Array.Copy(BitConverter.GetBytes(fourcc), 0, r, 0, 4);
                    Array.Copy(BitConverter.GetBytes(payloadBytes), 0, r, 4, 8);
                    Array.Copy(BitConverter.GetBytes(trackIdx), 0, r, 12, 4);
                }
                break;

            case RemoteCommandType.PlayPositionUpdate: {
                    /* fourcc              (4bytes)
                     * payload bytes == 12 (8bytes)
                     * state               (4bytes)
                     * trackIdx            (4bytes)
                     * positionMillisec    (4bytes)
                     */
                    r = new byte[24];
                    int fourcc = RemoteCommandTypeToFourCC(cmd);
                    long payloadBytes = 12;
                    Array.Copy(BitConverter.GetBytes(fourcc), 0, r, 0, 4);
                    Array.Copy(BitConverter.GetBytes(payloadBytes), 0, r, 4, 8);
                    Array.Copy(BitConverter.GetBytes((int)state), 0, r, 12, 4);
                    Array.Copy(BitConverter.GetBytes(trackIdx), 0, r, 16, 4);
                    Array.Copy(BitConverter.GetBytes(positionMillisec), 0, r, 20, 4);
                }
                break;

            case RemoteCommandType.PlaylistData: {
                /* 
                * "PLLS"
                * Number of payload bytes (int64)
                * 
                * playbackState    (int32)
                * 
                * Number of tracks (int32)
                * if (0 < number of tracks) {
                *   selected track (int32)
                * 
                *   Track0 duration millisec (int32)
                *   Track0 sampleRate        (int32)
                *   Track0 bitdepth          (int32)
                *   Track0 albumName bytes (int32)
                *   Track0 albumName (utf8 string)
                *   Track0 artistName bytes (int32)
                *   Track0 artistName (utf8 string)
                *   Track0 titleName bytes (int32)
                *   Track0 titleName (utf8 string)
                *   Track0 albumCoverArt bytes (int32)
                *   Track0 albumCoverArt (binary)
                * 
                *   Track1 
                *   ...
                * }
                */
                    List<byte[]> sendData = new List<byte[]>();

                    sendData.Add(BitConverter.GetBytes((int)state));

                    sendData.Add(BitConverter.GetBytes(playlist.Count));

                    if (0 < playlist.Count) {
                        sendData.Add(BitConverter.GetBytes(trackIdx));

                        int idx = 0;
                        foreach (var pl in playlist) {
                            sendData.Add(BitConverter.GetBytes(pl.durationMillsec));
                            sendData.Add(BitConverter.GetBytes(pl.sampleRate));
                            sendData.Add(BitConverter.GetBytes(pl.bitDepth));
                            AppendString(pl.albumName, ref sendData);
                            AppendString(pl.artistName, ref sendData);
                            AppendString(pl.titleName, ref sendData);
                            AppendByteArray(pl.albumCoverArt, ref sendData);
                            ++idx;
                        }
                    }
                    sendData.Insert(0, BitConverter.GetBytes(ByteArrayListBytes(sendData)));
                    sendData.Insert(0, BitConverter.GetBytes(FOURCC_PLAYLIST_DATA));

                    r = ByteArrayListToByteArray(sendData);
                } break;
            default:
                throw new NotImplementedException();
            }

            return r;
        }

        private static long ByteArrayListBytes(List<byte[]> a) {
            long bytes = 0;
            foreach (var item in a) {
                bytes += item.Length;
            }
            return bytes;
        }

        private static byte[] ByteArrayListToByteArray(List<byte[]> a) {
            int bytes = 0;
            foreach (var item in a) {
                bytes += item.Length;
            }

            var result = new byte[bytes];
            int offs = 0;
            foreach (var item in a) {
                Array.Copy(item, 0, result, offs, item.Length);
                offs += item.Length;
            }

            return result;
        }

        private static void AppendString(string s, ref List<byte[]> to) {
            if (s.Length == 0) {
                int v0 = 0;
                to.Add(BitConverter.GetBytes(v0));
                return;
            }

            var sBytes = System.Text.Encoding.UTF8.GetBytes(s);
            to.Add(BitConverter.GetBytes(sBytes.Length));
            to.Add(sBytes);
        }

        private static void AppendByteArray(byte[] b, ref List<byte[]> to) {
            if (b == null || b.Length == 0) {
                int v0 = 0;
                to.Add(BitConverter.GetBytes(v0));
                return;
            }

            to.Add(BitConverter.GetBytes(b.Length));
            to.Add(b);
        }
    };
}

File: PlayPcmWin/PPWRemote6/java/com/example/yamamoto2002/ppwremote6/RemoteCommand.kt
package com.example.yamamoto2002.ppwremote6

import android.util.Log
import java.io.IOException
import java.io.InputStream
import java.lang.System.arraycopy
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

/**
    @param trackIdx Track idx. Used on Play Pause and Seek.
 */
class RemoteCommand(var cmd : CommandType, var trackIdx : Int = 0) {
    enum class CommandType(val fourcc:Int) {
        PlaylistData(0x534c4c50),
        Exit(0x54495845),
        Pause(0x53554150),
        Stop(0x504f5453),
        PlayPositionUpdate(0x55504c50),

        Greetings(0x52575050),
        PlaylistWant(0x574c4c50),
        Play(0x59414c50),
        Seek(0x4b454553),
        SelectTrack(0x544c4553),
    }

    /// STATE_STOPPED STATE_PLAYING STATE_PAUSED
    var state : Int = 0

    /// Play position on track. Used on Play Pause and Seek.
    var positionMillisec: Int = 0

    /// used by Seek send
    var trackMillisec : Int = 0

    class PlayListItem {
        var durationMillsec: Int = 0
        var sampleRate : Int = 0
        var bitDepth : Int = 0
        var albumName: String = ""
        var artistName: String = ""
        var titleName: String = ""
        var albumCoverArt: ByteArray = ByteArray(0)
    }

    /// PlaylistData
    var playlist: MutableList<PlayListItem> = mutableListOf()

    companion object {
        const val VERSION: Long = 102

        const val STATE_STOPPED : Int = 0
        const val STATE_PLAYING : Int = 1
        const val STATE_PAUSED : Int = 2

        private const val TOO_LARGE : Long = 100 * 1000 * 1000
        private const val TOO_MANY_TRACKS : Long = 10000

        fun intToByteArray(v: Int): ByteArray {
            val b = ByteBuffer.allocate(4)
            b.order(ByteOrder.LITTLE_ENDIAN)
            b.putInt(v)
            return b.array()
        }

        fun longToByteArray(v: Long): ByteArray {
            val b = ByteBuffer.allocate(8)
            b.order(ByteOrder.LITTLE_ENDIAN)
            b.putLong(v)
            return b.array()
        }

        private fun byteArrayToInt(v :ByteArray) : Int {
            return ByteBuffer.wrap(v,0,4).order(java.nio.ByteOrder.LITTLE_ENDIAN).int
        }
        private fun byteArrayToLong(v :ByteArray) : Long {
            return ByteBuffer.wrap(v,0,8).order(java.nio.ByteOrder.LITTLE_ENDIAN).long
        }

        // ブロックして所望サイズのデータを受信する。
        private fun readBlocking(ins : InputStream, bytes : Int) : ByteArray {
            val r = ByteArray(bytes)
            var offs = 0
            while (offs < bytes) {
                if (0 == ins.available()) {
                    Thread.sleep(10)
                    //Log.i("RemoteCommand", "Sleep...")
                    continue
                }
                val wantBytes = bytes - offs
                val readBytes = ins.read(r, offs, wantBytes)
                if (readBytes == -1) {
                    throw IOException("readBlocking failed")
                }
                offs += readBytes
            }

            return r
        }

        private fun readInt(ins : InputStream) : Int {
            val buff = readBlocking(ins, 4)
            return byteArrayToInt(buff)
        }

        private fun readLong(ins : InputStream) : Long {
            val buff = readBlocking(ins, 8)
            return byteArrayToLong(buff)
        }

        private fun readByteArray(ins : InputStream) : ByteArray {
            val bytes = readInt(ins)
            if (bytes < 0 || TOO_LARGE < bytes) {
                throw IOException("String too large")
            }
            if (bytes == 0) {
                return ByteArray(0)
            }

            return readBlocking(ins, bytes)
        }

        private fun readUtf8(ins : InputStream) : String {
            val buff = readByteArray(ins)
            if (buff.isEmpty()) {
                return ""
            }
            return String(buff, StandardCharsets.UTF_8)
        }

        fun fromStream(ins : InputStream) : RemoteCommand {
            val cmd = RemoteCommand(CommandType.Exit)

            val fourcc = readInt(ins)
            val bytes = readLong(ins)

            if (bytes < 0 || TOO_LARGE < bytes) {
                return cmd
            }

            when (fourcc) {
                CommandType.PlaylistData.fourcc -> {
                    /*
                    * Number of payload bytes (int64)
                    *
                    * playbackState    (int32)
                    *
                    * Number of tracks (int32)
                    * if (0 < number of tracks) {
                    *   selected track (int32)
                    *
                    *   Track0 duration millisec (int32)
                    *   Track0 sampleRate        (int32)
                    *   Track0 bit depth         (int32)
                    *   Track0 albumName bytes (int32)
                    *   Track0 albumName (utf8 string)
                    *   Track0 artistName bytes (int32)
                    *   Track0 artistName (utf8 string)
                    *   Track0 titleName bytes (int32)
                    *   Track0 titleName (utf8 string)
                    *   Track0 albumCoverArt bytes (int32)
                    *   Track0 albumCoverArt (binary)
                    *
                    *   Track1
                    *   ...
                    * }
                    */

                    cmd.state = readInt(ins)
                    if (cmd.state < 0 || STATE_PAUSED < cmd.state) {
                        Log.e("RemoteCommand", "nTracks is out of range ${cmd.state}")
                        return cmd
                    }

                    val nTracks = readInt(ins)
                    if (nTracks < 0 || TOO_MANY_TRACKS < nTracks) {
                        Log.e("RemoteCommand", "nTracks is out of range $nTracks")
                        return cmd
                    }

                    if (nTracks == 0) {
                        cmd.cmd = RemoteCommand.CommandType.PlaylistData
                        return cmd
                    }

                    cmd.trackIdx = readInt(ins)
                    if (cmd.trackIdx < 0 || nTracks <= cmd.trackIdx) {
                        Log.i("RemoteCommand", "trackIdx is out of range ${cmd.trackIdx} fixed to 0")
                        cmd.trackIdx = 0
                    }

                    for (i in 0 until nTracks) {
                        //Log.i("RemoteCommand", "Reading track $i of $nTracks")

                        val p = RemoteCommand.PlayListItem()
                        p.durationMillsec = readInt(ins)
                        p.sampleRate = readInt(ins)
                        p.bitDepth = readInt(ins)
                        p.albumName = readUtf8(ins)
                        p.artistName = readUtf8(ins)
                        p.titleName = readUtf8(ins)
                        p.albumCoverArt = readByteArray(ins)
                       //Log.i("RemoteCommand", "p.albumCoverArt $i size = ${p.albumCoverArt.size}")
                        cmd.playlist.add(p)
                    }

                    cmd.cmd = CommandType.PlaylistData
                }
                CommandType.Pause.fourcc -> {
                    if (bytes != 4L) {
                        Log.e("RemoteCommand", "Pause bytes not equal to 4 $bytes")
                        return cmd
                    }
                    cmd.trackIdx = readInt(ins)
                    if (cmd.trackIdx < 0) {
                        Log.e("RemoteCommand", "trackIdx is negative ${cmd.trackIdx}")
                        return cmd
                    }
                    cmd.cmd = CommandType.Pause
                }
                CommandType.PlayPositionUpdate.fourcc -> {
                    if (bytes != 12L) {
                        Log.e("RemoteCommand", "PlayPositionUpdate bytes not equal to 12 $bytes")
                        return cmd
                    }
                    cmd.state = readInt(ins)
                    if (cmd.state < 0 || 2 < cmd.state) {
                        Log.e("RemoteCommand", "state is invalid ${cmd.state}")
                        return cmd
                    }
                    cmd.trackIdx = readInt(ins)
                    if (cmd.trackIdx < 0) {
                        Log.e("RemoteCommand", "trackIdx is negative ${cmd.trackIdx}")
                        return cmd
                    }
                    cmd.positionMillisec = readInt(ins)
                    if (cmd.positionMillisec < 0) {
                        Log.e("RemoteCommand", "positionMs is negative ${cmd.positionMillisec}")
                        return cmd
                    }
                    cmd.cmd = CommandType.PlayPositionUpdate
                }
                CommandType.Exit.fourcc -> {
                    if (bytes != 0L) {
                        Log.e("RemoteCommand", "Exit bytes not equal to 0 $bytes")
                        return cmd
                    }
                }
                CommandType.Stop.fourcc -> {
                    if (bytes != 0L) {
                        Log.e("RemoteCommand", "Stop bytes not equal to 0 $bytes")
                        return cmd
                    }
                    cmd.cmd = CommandType.Stop
                }
                else -> {
                    Log.i("RemoteCommand", "Unknown fourcc $fourcc")
                }
            }
            return cmd
        }
    }

    private fun countByteArrayListTotalBytes(a : List<ByteArray>) : Int {
        var bytes = 0
        a.onEach { bytes += it.size }

        return bytes
    }

    private fun byteArrayListToByteArray(a : List<ByteArray>) : ByteArray {
        val bytes = countByteArrayListTotalBytes(a)

        val r = ByteArray(bytes)
        var offs = 0
        a.onEach {
            arraycopy(it,0,r,offs,it.size)
            offs += it.size
        }

        return r
    }

    fun toByteArray() : ByteArray {
        val r = mutableListOf<ByteArray>()

        when (cmd) {
            CommandType.Greetings -> {
                /*
                  "Greetings"
                  size    = 8   (4bytes)
                  version = 100 (8bytes)
                 */
                r.add(longToByteArray(VERSION))
            }
            CommandType.PlaylistWant -> {
                // no payload
            }
            CommandType.Exit -> {
                // no payload
            }
            CommandType.Pause -> {
                // no payload
            }
            CommandType.Play -> {
                /*
                   "PLAY"
                   trackIdx (4bytes)
                 */
                r.add(intToByteArray(trackIdx))
            }
            CommandType.Seek -> {
                /*
                    total 8bytes
                    positionMillisec (4bytes)
                    trackMillisec (4bytes)
                 */
                r.add(intToByteArray(positionMillisec))
                r.add(intToByteArray(trackMillisec))
            }
            else -> {
                throw NotImplementedError("RemoteCommand.toByteArray")
            }
        }

        // calc payload size
        val payloadBytes = countByteArrayListTotalBytes(r)

        // insert header and payload bytes in the front
        r.add(0, longToByteArray(payloadBytes.toLong()))
        r.add(0, intToByteArray(cmd.fourcc))

        return byteArrayListToByteArray(r)
    }

}